### DISCUSSION TITLE
some question about the server side code

### DISCUSSION THREAD

**Discussion by gengtsh** [carla-simulator/carla#4923] [Q&A] [answered]
hi,
can someone point out where the server side code is? 
in the client side, i think detail::Client::Pimpl call rpc::Client, and then rpc::Client call ::rpc::client , finally, ::rpc::client invoke client.call() function to wait for the response of ::rpc::server. The rpc::Server.AsyncRun() start the work thread, in other words, it start the ::rpc::server. but who call rpc::Server.AsyncRun()? 
how to connect ::rpc::server with unreal engine? I'm really have a problem to build up the caller graph in server side. 
i'm also very confused about the the relationship between  folder _carla\Unreal\CarlaUE4\Plugins\Carla\Source\Carla_ and  folder _carla\LibCarla\source\carla_ 

when running make launch command, UE4Ediotor is launched, how the connections among the Client, Server and UE4Ediotor are built up  confuses me quite a long time. can someone help me out? Thanks a lot.

**Comment by ynlch** [comment, marked as answer]
carla\Unreal\CarlaUE4\Plugins\Carla\Source\Carla, all these files are plugin code, run from UE4, which launch with CarlaUE4.sh
 
carla\LibCarla\source\carla is used both by libcarla_client.a(linked into your client app) and libcarla_server.a(linked into plugin), to support c/s streaming

have a look LibCarla/cmake/client/CMakeLists.txt and server/CMakeLists.txt, it will be clear
